Sources/System/FilePath/FilePathComponentView.swift

```diff
@@ -0,0 +1,220 @@
+  /// standard `RangeReplaceableCollection` algorithms for modifying the
+  /// file path using component or range of components granularity.
+  ///
+  /// Example:
+  ///
+  ///     var path: FilePath = "/./home/./username/scripts/./tree"
+  ///     let scriptIdx = path.components.lastIndex(of: "scripts")!
+  ///     path.components.insert("bin", at: scriptIdx)
+  ///     // path is "/./home/./username/bin/scripts/./tree"
+  ///
+  ///     path.components.removeAll { $0.kind == .currentDirectory }
+  ///     // path is "/home/username/bin/scripts/tree"
+  ///
+  public struct ComponentView {
+    internal var _path: FilePath
+    internal var _start: SystemString.Index
+
+    internal init(_ path: FilePath) {
+      self._path = path
+      self._start = path._relativeStart
+      _invariantCheck()
+    }
+  }
+
+  /// View the non-root components that make up this path.
+  public var components: ComponentView {
+    __consuming get { ComponentView(self) }
+    _modify {
+      // RRC's empty init means that we cann't guarantee that the yielded
+      // view will restore our root. So copy it out first.
+      //
+      // TODO(perf): Small-form root (especially on Unix). Have Root
+      // always copy out (not worth ref counting). Make sure that we're
+      // not needlessly sliding values around or triggering a COW
+      let rootStr = self.root?._systemString ?? SystemString()
+      var comp = ComponentView(self)
+      self = FilePath()
+      defer {
+        self = comp._path
+
+        if !rootStr.isEmpty {
+          if let r = self.root {
+            // Roots can be forgotten but never altered
+            assert(r._slice.elementsEqual(rootStr))
+          }
+          // TODO: conditional on it having changed?
+          self.root = Root(rootStr)
+        }
+      }
+      yield &comp
+    }
+  }
+}
+
+// @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension FilePath.ComponentView: BidirectionalCollection {
+  public typealias Element = FilePath.Component
+  public struct Index: Comparable, Hashable {
+    internal typealias Storage = SystemString.Index
+
+    internal var _storage: Storage
+
+    public static func < (lhs: Self, rhs: Self) -> Bool {
+      lhs._storage < rhs._storage
+    }
+
+    fileprivate init(_ idx: Storage) {
+      self._storage = idx
+    }
+  }
+
+  public var startIndex: Index { Index(_start) }
+  public var endIndex: Index { Index(_path._storage.endIndex) }
+
+  public func index(after i: Index) -> Index {
+    return Index(_path._parseComponent(startingAt: i._storage).nextStart)
+  }
+
+  public func index(before i: Index) -> Index {
+    Index(_path._parseComponent(priorTo: i._storage).lowerBound)
+  }
+
+  public subscript(position: Index) -> FilePath.Component {
+    let end = _path._parseComponent(startingAt: position._storage).componentEnd
+    return FilePath.Component(_path, position._storage ..< end)
```

Comment:
Is it worth making ComponentView.Index store a Range so subscript is O(1)?

Reply:
It is amortized constant, as in a subscript over every index will be a total of O(n). We could store the range, but that would complicate `startIndex` and/or the initializer for ComponentView. If we have an `unknown` representation for the length, that further complicates other logic. `subscript` is not the only operation to optimize for, as many operations (especially RRC operations) only advance indices or slice the collection.

This "parse" is just scanning bytes looking for the `/` byte (or `\` on Windows). I don't think it's worth spending the complexity budget on, since these types are resilient, these operations are not inlinable, and the scanned byte range is loaded at the end (with a retain, at least until we do small-form for small components).

Reply:
It's good we have the flexibility to add this in the future.

Computing the Range in the ComponentView initializer seems like the right direction to me. It's the direction we've taken with the [`ChunkedByCount`](https://github.com/apple/swift-algorithms/blob/main/Sources/Algorithms/Chunked.swift#L285-L290)  and [`SlidingWindows`](https://github.com/apple/swift-algorithms/blob/main/Sources/Algorithms/SlidingWindows.swift#L45) collections from the Algorithms package.

Reply:
By "resilient" I wasn't referring to the ability to add it later (though that's true), I was referring to the performance characteristics of the code. Inside the call, subscript does something computers are really good at (linear scan within a cache line for data that will be loaded anyways), after having been called indirectly and returning its result indirectly.

From a performance perspective, I don't think `ChunkedByCount` and `SlidingWindows` are analogous, as their collection status is primarily for the purpose of yielding elements. They're not RRC, for example.